### ISSUE TITLE
Package lock, optional dependencies and cross-platform issue

### ISSUE THREAD

**Issue by soundstep** [npm/npm#18202]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

It took me 2 days to debug an issue.
I thought I'd open one here even though it is not exactly a NPM issue, it might need a bit of thinking.

We are using on a repo the `pact-node` npm module, which has a `pact-mock-service-npm` dependency, This `pact-mock-service-npm` repo is using the `optionalDependencies` field to install a per-OS dependency:

```
"optionalDependencies": {
    "@pact-foundation/pact-mock-service-linux-ia32": "2.1.0",
    "@pact-foundation/pact-mock-service-linux-x64": "2.1.0",
    "@pact-foundation/pact-mock-service-win32": "2.1.0",
    "@pact-foundation/pact-mock-service-darwin": "2.1.0"
  }
```

I just recently switched to node v8.3.0, and my mistake was to include the `package-lock.json` in the repo. This had the effect of storing a darwin dependency, as we develop on MacOS. This is how the `package-lock.json` looks like:

```
"dependencies": {
        "@pact-foundation/pact-mock-service-darwin": "0.10.2"
```

This made a job on our pipeline fails as it run on a centos linux x64.

I just wanted to share the problem I had in case there is anything to improve on the npm install side to avoid this kind of clash with optional dependencies.

More info there:

https://github.com/pact-foundation/pact-node/issues/42

#### How can the CLI team reproduce the problem?

1. Create a package.json using '@pact-foundation/pact-node'
2. Execute `npm install` with node v8.3.0 on a certain OS (let's say MacOS)
3. Push the package.json and package-lock.json a repo
4. Clone this repo on another OS (let's say linux)
5. Execute `npm install` on the other OS (linux)

Result: the `package-lock.json` file contains information execute from a MacOS machine, which will make the wrong dependency install if it run on another OS.

This is not an issue as the installation does not fail. but the concept of the package-lock.json and `optionalDependencies`, if they are meant to be used to do per-OS installation, clashes a bit.

Any thoughts?

### supporting information:

 - `npm -v` prints: 5.3.0
 - `node -v` prints: 8.3.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: cross platform issue
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

#### Is the absence of this feature blocking you or your team? If so, how?

Not blocking, but we will now remove `package-lock.json` files from our repos.

**Comment by mboudreau** [comment]
I'm one of the developers of pact-node and willing to help debug this issue further if I'm needed.  Our product really hinges on optional dependencies, each of which specify the OS and architecture, so the end result should be only a single dependency is installed.  We also have a postinstall script that confirms this in case it comes along an unknown OS or architecture and warns the developer before using our product.

Cheers.

**Comment by fizker** [comment]
I have a similar issue with `fsevents` being installed from a mac, and failing on windows. To make it worse, package-lock.json is updated by the windows machine to remove `fsevents`, which not only gives a lot of noise, but also goes against the documentation (https://docs.npmjs.com/files/package-lock.json#optional): `All optional dependencies should be included even if they're uninstallable on the current platform`

**Comment by xfh** [comment]
@fizker I am also facing this issue with fsevents. In my case fsevents is pulled in by chokidar. Chokidar actually declars fsevents under `dependencies` and under `optionalDependencies`. Maybe that's why package-lock does not handle the optional dependency correctly?

**Comment by fizker** [comment]
@xfh Hm, I also get `fsevents` from `chokidar` (which comes from `babel-cli`). But upon investigation, your statement is not actually true. 

While the package.json for the installed `chokidar` does have fsevents in both `dependencies` and `optionalDependencies`, both https://github.com/paulmillr/chokidar/blob/master/package.json and the package.json from https://registry.npmjs.org/chokidar/-/chokidar-1.7.0.tgz (the link that my installed `chokidar` has for the `_resolved` key) have `fsevents` only in `optionalDependencies`.

The faulty doubling of `fsevents` must then come from npm during the `npm install` routine.

**Comment by xfh** [comment]
@fizker you got me, I only checked the package.json from the installed `chockidar` version.
So it seems like an npm issue after all. Thanks for the correction.

**Comment by felixfbecker** [comment]
This is preventing big cross-platform open source projects like VS Code from updating https://github.com/Microsoft/vscode/issues/30134

**Comment by danepowell** [comment]
I'm also having this problem, due to fsevents being pulled in by chokidar 1.6.1. This is in spite of the fact that fsevents is clearly only listed as an optional dependency of chokidar: https://github.com/paulmillr/chokidar/blob/1.6.1/package.json#L43

At some point, a bug in npm is turning this into a hard dependency in package-lock.json.

**Comment by felixfbecker** [comment]
@danepowell having it in package-lock.json is actually the expected behaviour, so when you run `npm install` on a mac it should pull in that locked version. But when running `npm install` on Linux or Windows, it should then just ignore it.

**Comment by danepowell** [comment]
I guess the problem I'm seeing is that the contents of `package-lock.json` change depending on what environment you run `npm install` on. It constantly pollutes Git history and pull requests with irrelevant changes.

Also, while you might be right that this is benign if you _first_ run npm install on Mac and _then_ on Linux, I'm not sure that the opposite is true. If you run npm install on Linux first (so package-lock.json is missing fsevents), and then later on Mac, what happens then? Presumably it's either missing the dependency, or it overwrites package-lock.json to add it. Either way, it's a bad UX.

**Comment by felixfbecker** [comment]
Yeah, that's the bug. It should always add `optionalDependencies` to package-lock.json and never remove them because being on a different platform, no matter whether the dependencies were actually installed or not.

**Comment by mightyiam** [comment]
Seems like a duplicate of #17722.
